README.md

```diff
@@ -0,0 +1,812 @@
+
+Sets the specified preset in the context.
+
+Returns whether the operation succeeded.
+
+#### bitrate()
+
+Returns the context's current bitrate.
+
+#### setBitrate()
+
+Sets the specified bitrate in the context.
+
+Special values are:
+
+- `0` to set the default value.
+- `infinite32` to set the maximum value supported.
+
+Returns whether the operation succeeded.
+
+#### inDTX()
+
+Returns whether the context is in DTX mode.
+
+It is when last frame was either ignored or a comfort noise frame.
+
+#### phaseInversion()
+
+Returns whether phase inversion is enabled.
+
+#### togglePhaseInversion()
+
+Enables or disables phase inversion, depending on the passed flag.
+
+Returns whether the operation succeeded.
+
+#### vbr()
+
+Returns whether variable bitrate is enabled.
+
+#### toggleVBR()
+
+Enables or disables variable bitrate, depending on the passed flag.
+
+Returns whether the operation succeeded.
+
+### Pack
+
+Used to serialize and deserialize Protobuf messages.
+
+Its internal buffer is a vector of bytes and can hold a serialized message, ready to be sent.  
+The first 6 bytes of the buffer represent the header as a `NetHeader` object.
+
+An object can be constructed through:
+
+- Copy
+- Move
+- `Message` (native or Protobuf's)
+- `NetHeader` (+ data size if UDP)
+
+Allocating extra space for the internal buffer is possible and it's what `Peer` does internally for UDP, to improve efficiency.  
+The buffer size is set to 1024 bytes, which is the formal maximum for our protocol, and the `Pack` object is recycled for every packet received.
+
+#### operator()
+
+Deserializes the held message into the `Message` object that is passed via reference.
+
+`dataSize` is only required to be specified when the vector of bytes is not fully occupied.  
+The Protobuf library's deserializer fails if it encounters extra bytes that are not part of the message.
+
+Returns whether the operation succeeded.
+
+#### buf()
+
+Returns a reference to the full vector of bytes.
+
+This is generally what you want to use when sending messages.
+
+The pointed data is read-only if the object is `const`.
+
+#### data()
+
+Returns a reference to the vector of bytes, after skipping the `NetHeader` object.
+
+The pointed data is read-only if the object is `const`.
+
+#### header()
+
+Returns a reference to the `NetHeader` object.
+
+The pointed data is read-only if the object is `const`.
+
+#### type()
+
+Returns the message type specified in the `NetHeader` object as a `Type` enum value.
+
+### Peer
+
+This is the networking core of your client or server.
+
+#### connect()
+
+Establishes a connection (TCP) to the specified endpoint (`peerEndpoint`).
+
+`endpoint` can be filled in order to use a specific source endpoint.  
+Especially useful when you have multiple Internet connections and want to force a specific one.
+
+Returns an [`std::pair`](https://en.cppreference.com/w/cpp/utility/pair) containing the result code and a socket handle (also called "file descriptor").  
```

Comment:
This seems odd. We are writing C++ and thus errors should be communicated by means of exceptions and not via result codes.

Reply:
Honestly, I've ever only used exceptions when strictly required, either by the standard library or an external one.

The "modern" way would be to use [`std::expected`](https://en.cppreference.com/w/cpp/utility/expected), as it's officially introduced in C++23.
A potential alternative is [`Boost.Outcome`](https://www.boost.org/doc/libs/1_80_0/libs/outcome/doc/html/index.html).

Reply:
Exceptions are still the best way of error handling, imo as they can't be ignored. They have to be handled or else the program terminates. It is unfortunate that C++ compilers don't warn you about potentially unhandled exceptions, though :shrug: 

Using monads is not necessarily the "modern" way of doing things. It's just another way.
However, since we don't have monads yet anyway, I would argue that exceptions are the best thing we got for the moment (certainly much better than error codes)

Reply:
I would argue it also depends on the way they're presented, in this library the behavior is very consistent.

On the other hand, unhandled exceptions generally trigger an automatic backtrace, which may be desiderable.

Reply:
Even if the error codes are very consistently returned, it is just so easy to forget to check them. I've seen and also experienced this so many times and the result of that is a program that doesn't do error checking, which more often than not will lead to the program crashing somewhere down the line and then it will be much more difficult to figure out where the error came from. With exceptions, you get a crash right when and where the error appears (if you don't handle it).